src/app/rpmostree-dbus-helpers.c

```diff
@@ -1154,3 +1158,124 @@ rpmostree_update_deployment (RPMOSTreeOS  *os_proxy,
                                                    cancellable,
                                                    error);
 }
+
+static void
+append_to_summary (GString     *summary,
+                   const char  *type,
+                   guint        n)
+{
+  if (n == 0)
+    return;
+  if (summary->len > 0)
+    g_string_append (summary, ", ");
+  g_string_append_printf (summary, "%u %s", n, type);
+}
+
+static const char*
+ondiskstate_to_str (RpmOstreeUpdateOnDiskStates on_disk_state)
+{
+  switch (on_disk_state)
+    {
+    case RPM_OSTREE_UPDATE_MISSING:
+      return "not downloaded";
+    case RPM_OSTREE_UPDATE_DOWNLOADED:
+      return "downloaded";
+    default:
+      return "unknown";
+    }
+}
+
+/* this is used by both `status` and `upgrade --check/--preview` */
+gboolean
+rpmostree_print_cached_update (GVariant         *cached_update,
+                               gboolean          verbose,
+                               GCancellable     *cancellable,
+                               GError          **error)
+{
+  g_auto(GVariantDict) dict;
+  g_variant_dict_init (&dict, cached_update);
+
+  /* let's just extract 📤 all the info ahead of time */
+
+  const char *checksum;
+  g_assert (g_variant_dict_lookup (&dict, "checksum", "&s", &checksum));
+
+  const char *version;
+  if (!g_variant_dict_lookup (&dict, "version", "&s", &version))
+    version= NULL;
+
+  g_autofree char *timestamp = NULL;
+  { guint64 t;
+    if (!g_variant_dict_lookup (&dict, "timestamp", "t", &t))
+      t = 0;
+    timestamp = rpmostree_timestamp_str_from_unix_utc (t);
+  }
+
+  gboolean gpg_enabled;
+  if (!g_variant_dict_lookup (&dict, "gpg-enabled", "b", &gpg_enabled))
+    gpg_enabled = FALSE;
+
+  g_autoptr(GVariant) signatures =
+    g_variant_dict_lookup_value (&dict, "signatures", G_VARIANT_TYPE ("av"));
+
+  gboolean is_new_checksum;
+  g_assert (g_variant_dict_lookup (&dict, "is-new-checksum", "b", &is_new_checksum));
+
+  g_autoptr(GVariant) rpm_diff =
+    g_variant_dict_lookup_value (&dict, "rpm-diff", G_VARIANT_TYPE ("a{sv}"));
+
+  RpmOstreeUpdateOnDiskStates on_disk_state;
+  g_assert (g_variant_dict_lookup (&dict, "on-disk-state", "u", &on_disk_state));
+
+  /* and now we can print 🖨️ things! */
+
+  g_print ("Available update:\n");
+
+  /* add the long keys here */
+  const guint max_key_len = MAX (strlen ("OnDiskState"),
+                                 strlen ("GPGSignature"));
+
+  if (is_new_checksum)
+    {
+      rpmostree_print_timestamp_version (version, timestamp, max_key_len);
+      rpmostree_print_kv ("Commit", max_key_len, checksum);
+      if (gpg_enabled)
+        rpmostree_print_gpg_info (signatures, verbose, max_key_len);
+      /* XXX: for now on-disk-state only represents ostree commit state; once it also
+       * includes layered rpm on-disk state, it would make sense to print this
+       * unconditionally; see related XXX in deployment-utils.c */
```

Comment:
It basically referred to whether the update was already downloaded or not. E.g. in `download` mode, we would expect a cached update to already be available locally. Since we dropped `download` mode for now, I also removed this key. It would probably be better served by a `DownloadSize` style key anyway, where `0` means "already downloaded".